Queue strategy top-down-level-by-level

surya4 opened the pull request:
https://aswattel.atlassian.net/browse/ZIWO-5632

Commit:
new queue startegy top-down-level-by-level

Commit:
gc comment

Commit:
startegy name changed from top-down to top-down-level-by-level

Commit:
keywords modified

royalpinto commented on src/mod/applications/mod_callcenter/mod_callcenter.c:
```diff
@@ -2452,6 +2454,27 @@ static int agents_callback(void *pArg, int argc, char **argv, char **columnNames
+						switch_channel_set_variable(member_channel, "cc_offered_level_agent", offered_agent_list);
+						switch_channel_set_variable(member_channel, "cc_last_agent_tier_level", agent_tier_level);
+					}
+						switch_core_session_rwunlock(member_session);
```
@surya4 indent back.

surya4 replied:
this is fine in code

royalpinto commented on src/mod/applications/mod_callcenter/mod_callcenter.c:
```diff
@@ -40,6 +40,7 @@
 #define CC_SQLITE_DB_NAME "callcenter"
 #define CC_APP_KEY "mod_callcenter"
 
+#define CC_OFFERED_AGENT_LIST_SIZE 10000
```
@surya4  is it not too big?

surya4 replied:
done @royalpinto reduced it to 2048 now so it can handle approx 128 agents in one level, if more required we can increase list size - '000000@default' -> 1+6+1+7+1=16 characters

royalpinto commented on src/mod/applications/mod_callcenter/mod_callcenter.c:
```diff
@@ -2452,6 +2454,27 @@ static int agents_callback(void *pArg, int argc, char **argv, char **columnNames
+
+						switch_channel_t *member_channel = switch_core_session_get_channel(member_session);
+
+						if ((offered_level_agent = switch_channel_get_variable(member_channel, "cc_offered_level_agent"))) {
+							snprintf(offered_agent_list, sizeof offered_agent_list, "%s,'%s'", offered_level_agent, h->agent_name);
+						} else {
+							snprintf(offered_agent_list, sizeof offered_agent_list, "'%s'", h->agent_name);
+						}
```
@surya4 can it also clear the offered list if you are jumping to the next level. to optimize the list size.

royalpinto replied:
check if previous and new offering level is same. if not clear and add instead of appending.

surya4 replied:
done @royalpinto

royalpinto reviewed (changes requested):
Good one @surya4, minor improvements if possible to tune.

Commit:
review fix:
1. reduce list size from 10000 to 2048
2. clear agent list when level is changed
3. modified variable names for more readability

wowi42 commented on src/mod/applications/mod_callcenter/mod_callcenter.c:
```diff
@@ -40,6 +40,7 @@
 #define CC_SQLITE_DB_NAME "callcenter"
 #define CC_APP_KEY "mod_callcenter"
 
+#define CC_OFFERED_AGENT_LIST_SIZE 2048
```
why 2048?

surya4 replied:
@wowi42 
this is used to maintain a string array of agent names at each level that will be used to ignore after they are offered once so, the ringing call can be offered to other available agents at this level.

example -> agents available at level 1:  cclogin1@default, cclogin2@default, cclogin3@default

say call comes and randomly offered to agent -> cclogin2@default, if he does not pick then call will go to either of the left agents and this agent will go to ignore list in below query to pick next available agent - 

query to fetch available agent -> 
`SELECT instance_id, name, status, contact, no_answer_count, max_no_answer, reject_delay_time, busy_delay_time, no_answer_delay_time, tiers.state, agents.last_bridge_end, agents.wrap_up_time, agents.state, agents.ready_time, tiers.position as tiers_position, tiers.level as tiers_level, agents.type, agents.uuid, external_calls_count, agents.last_offered_call as agents_last_offered_call FROM agents LEFT JOIN tiers ON (agents.name = tiers.agent) WHERE tiers.queue = 'abraKaDabra@default' AND (agents.status = 'Available' OR agents.status = 'On Break' OR agents.status = 'Available (On Demand)') AND tiers.level >= 2 AND name NOT IN ('0002@default') ORDER BY tiers_level ASC, random(), agents_last_offered_call`


the max an agent name can go is -> 16 characters ('000000@default')
so at each level, we can server ~110 agents in the ignore list

Commit:
variable name changes and indent fix after review

Commit:
changes to debug agent busy status

Commit:
remove busy user status from query
